AI Logic : Handling All Card Combinations

jduong09 opened the pull request:
Partially resolves #22 

## Changes/Additions
- Added logic so AI can be given current turn combo (single/high card | pair | three of a kind | four of a kind | sequence | double sequence)
- If AI has multiple possible plays for a turn combo, AI will choose lowest value combo that beats the middle pile value

## Hot To Test
- Checkout to `issue-22--ai-logic` and pull changes with `git pull`
- Navigate to `localhost:3000` in your browser
- Shuffle deck, the first turn will be user. Choose combination type you want to test and play valid turn.
- CPU will try to play valid turn and pass if there is no valid turn.
- Use console logs to verify correct turn and follow along with middle pile as well.
- When cycle ends, refresh browser to continue testing.

ziggysauce commented on components/game/game.js:
```diff
@@ -238,6 +238,7 @@ const Game = () => {
     );
   });
 
+  console.log(hands);
```
remove? unless you want to keep this intentionally to see for testing

jduong09 replied:
Removed.

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+'clubs': '♣',
+*/
+
+const createDuplicatesObject = (array) => {
```
move this to the top of the file since it's reference before you've instantiated it here

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+/* AI logic to play a double */
+// Need to handle logic if there are 3/4 of a kind
+// This leads to different combinations of doubles.
```
it's a little cleaner to keep your comment styles consistent. For single lines i recommend just the double forward slashes. for multiple lines, the large /* */ comment style is preferred.

ex:
```javascript
// This is a single comment
const foo = bar;
```

```javascript
/*
 * This is a multi-line comment
 * So we prefer to use this style
 */
const foo = bar;
```

jduong09 replied:
I'll keep this in mind, I removed the multi-lines, but going forward I gotchu.

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+const hand = [ { number: 12, suite: 'hearts', value: 36, selected: false, copy: 3 }, { number: 12, suite: 'hearts', value: 36, selected: false, copy: 4 }, { number: 6, suite: 'hearts', value: 16, selected: false },  { number: 14, suite: 'hearts', value: 48, selected: false }, { number: 12, suite: 'hearts', value: 40, selected: false, copy: 1 }, { number: 8, suite: 'diamonds', value: 23, selected: ...
+// const hand = [ { number: 11, suite: 'hearts', value: 36, selected: false }, { number: 6, suite: 'hearts', value: 16, selected: false },  { number: 14, suite: 'hearts', value: 48, selected: false }, { number: 12, suite: 'hearts', value: 40, selected: false }, { number: 8, suite: 'diamonds', value: 23, selected: false }, { number: 9, suite: 'hearts', value: 28, selected: false }, { number: 15, s ...
+
+console.log(quartet(hand));
```
remove?

jduong09 replied:
Removed.

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+}
+
+//const hand = [ { number: 11, suite: 'hearts', value: 36, selected: false }, { number: 6, suite: 'hearts', value: 16, selected: false },  { number: 14, suite: 'hearts', value: 48, selected: false }, { number: 12, suite: 'hearts', value: 40, selected: false, copy: 1 }, { number: 8, suite: 'diamonds', value: 23, selected: false }, { number: 9, suite: 'hearts', value: 28, selected: false }, { numbe ...
+const hand = [ { number: 12, suite: 'hearts', value: 36, selected: false, copy: 3 }, { number: 12, suite: 'hearts', value: 36, selected: false, copy: 4 }, { number: 6, suite: 'hearts', value: 16, selected: false },  { number: 14, suite: 'hearts', value: 48, selected: false }, { number: 12, suite: 'hearts', value: 40, selected: false, copy: 1 }, { number: 8, suite: 'diamonds', value: 23, selected: ...
```
for readability purposes, i suggest you format this to use new lines

ex:
```javascript
const hand = [
  { number: 12, suite: 'hearts', value: 36, selected: false, copy: 3},
  { number: 12, suite: 'hearts', value: 36, selected: false, copy: 4 },
  ...
]
```

jduong09 replied:
Removed, testing is done through localhost.

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+          combinations.push([dupesObject[value][i], dupesObject[value][j]]);
+        }
+      }
+    }
```
what happens if `dupesObject[value].length < 2`? I assume we'll need that logic wherever this is called to avoid running this function with no output. But i guess qorst case scenario it'll just result in an empty array for `combinations` which is fine.

jduong09 replied:
You are correct, what happens is an empty array is sent, and in Game Component, if empty array is received then ai will pass turn.

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+    }
+  }
+  return combinations;
+}
+/* AI logic to play a triple */
```
for cleaner readability, add a line break between functions

```suggestion
}

/* AI logic to play a triple */
```

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+    }
+  }
+  return combinations;
+}
+/* AI logic to play a quartet */
```
same comment here

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+const merge = (left, right) => {
+  let sortedArr = [];
+
+  while (left.length && right.length) {
```
hmm, am i missing something? in theory `left` and `right` could have a length, thus this `while` loop will never break?

jduong09 replied:
In the while loop, the array is being reduced until it is empty by using left.shift() and right.shift()

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+  return merge(left, right);
+}
+
+/*
+'hearts': '♥',
+'diamonds': '♦',
+'spades': '♠',
+'clubs': '♣',
+*/
```
what's this for?

jduong09 replied:
Comment to help with my testing while coding this PR, removed.

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+
+  for (let i = 0; i < array.length; i++) {
+    const number = array[i].number;
+    if (!values[number]) {
+      values[number] = [array[i]];
+    } else {
+      values[number].push(array[i]);
+    }
```
standard eslint rules prefer the if to not be a negated expression. thus you'd want to write it like

```suggestion
    if (values[number]) {
      values[number].push(array[i]);
    } else {
      values[number] = [array[i]];
    }
```

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,184 @@
+  const sortedHand = mergeSort(hand);
+  const totalSequences = [];
+
+  for (let i = 0; i < sortedHand.length; i++) {
```
i'm going to need to take another look at this. and probably also at the functions above as well. you have a lot of for loops nested within for loops, and i think some of these can probably be re-written as reduce functions to make this a bit simpler to read and possibly even run faster.

jduong09 replied:
Leaving this comment so you can take a look again. I'll take a look when I have free time.

ziggysauce reviewed (changes requested):
niice mostly working for me. i did notice however when going through the cycles, the selection disappears and isn't visible until you click on a specific combination type

<img width="927" alt="Screen Shot 2024-11-08 at 9 50 25 PM" src="https://github.com/user-attachments/assets/b52d469c-6a08-4736-a8e5-bc18ac248e5f">
<img width="310" alt="Screen Shot 2024-11-08 at 9 50 31 PM" src="https://github.com/user-attachments/assets/ad323826-c289-49c2-b42c-f73dce277097">

ziggysauce commented on components/utilities/ai.js:
```diff
@@ -0,0 +1,273 @@
+    let startingNumber = sortedHand[i].number;
+    let sequences = [];
+
+    // if starting iteration does not containn two of its number, continue to next iteration.
```
```suggestion
    // if starting iteration does not contain two of its number, continue to next iteration.
```

jduong09 commented:
> niice mostly working for me. i did notice however when going through the cycles, the selection disappears and isn't visible until you click on a specific combination type
> 
> <img alt="Screen Shot 2024-11-08 at 9 50 25 PM" width="927" src="https://private-user-images.githubusercontent.com/25065329/384572018-b52d469c-6a08-4736-a8e5-bc18ac248e5f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0ODA3NDMsIm5iZiI6MTczMjQ4MDQ0MywicGF0aCI6Ii8yNTA2NTMyOS8zODQ1NzIwMTgtYjUyZDQ2OWMtNmEwOC00NzM2LWE4ZTUtYmMxOGFjMjQ4ZTVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI0VDIwMzQwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyNWRmN2Y0NzE2MTYyYzNlZTFkZjlkNTA5MmRjN2MwYjI1NTVhODM1MzViNjkwMDkxNzhmZTk3MDQzY2VjNDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Iyh5anJ99f9J3NDnm6lLLQOdA5LT9DdomJMPLxypBms"> <img alt="Screen Shot 2024-11-08 at 9 50 31 PM" width="310" src="https://private-user-images.githubusercontent.com/25065329/384572021-ad323826-c289-49c2-b42c-f73dce277097.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0ODA3NDMsIm5iZiI6MTczMjQ4MDQ0MywicGF0aCI6Ii8yNTA2NTMyOS8zODQ1NzIwMjEtYWQzMjM4MjYtYzI4OS00OWMyLWI0MmMtZjczZGNlMjc3MDk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI0VDIwMzQwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkNDI1YWU2MzU4MjI4ZDQ2MzE4ZTcyOThmZmI2M2JiM2U4Mjc5OWNhM2FhYjhiMjhiOTJkMDhjZTU5ZWY0OTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K0BSHNtNitgcSXoerAaNQ77kBtJnExz3i2x6H2JANWM">

Yep, I had it like this, to "reset" the next turn. In the previous version, at the end of the current cycle, the cycles reset and it stays as "single", but in the game's states it was reset back to "". This current change, makes it so the user understands that they need to re-set the combination type back to something.

jduong09 commented:
## 11/24/24 Update

- Added state for turn messages.
- Fixed error occurring when user finalized turn when combination type is not set.

### Comments
- I added a state to better understand when the user needs to set a combination type, especially after a turn cycle has been completed and we reset the round. This is probably temporary for this PR, as we will have better UI in the future for game flow.

ziggysauce commented:
gotcha. i think what i was running into was the combinations not sticking through the round, but maybe i'm mis-remembering. i'll check again.

ziggysauce reviewed (commented):
i like the addition of turn messages, even if just for testing rn with the UI. it may make sense to make this more of an alert or notification in actual game play. so maybe it can be sticky. the UI is still up for grabs

jduong09 commented:
Discussion on Friday, December 6th. Approved.
